File: ts/src/engine.ts
/** Reconciliation cascade orchestrator.
 *
 * Order is a correctness requirement:
 *   Pass 0 duplicates -> 1 exact ref -> 2 normalised ref
 *   -> 3 ref-match/amount-differ -> 4 amount+date -> 5 subset sums
 *   -> 6 classify residuals -> bridge.
 *
 * Pass 0 must run before any matching: if matching ran first, one of a
 * duplicate pair would be consumed by a match and the duplicate would become
 * undetectable. */
import { buildBridge } from "./bridge";
import { BUCKET_ORDER } from "./models";
import type {
  Finding, LedgerLine, Match, ReconcileResult, StatementLine,
} from "./models";
import {
  pass0Duplicates, pass1ExactRef, pass2NormalisedRef, pass3RefAmountDiffer,
  pass4AmountDate, pass5SubsetSums, pass6Classify,
} from "./passes";

export interface Config {
  duplicateWindowDays: number;
  duplicateTightDays: number;
  amountDateWindowDays: number;
  timingDays: number;
  subsetMaxSize: number;
}

export const DEFAULT_CONFIG: Config = {
  duplicateWindowDays: 30,
  duplicateTightDays: 7,
  amountDateWindowDays: 5,
  timingDays: 3,
  subsetMaxSize: 6,
};

function cmp(a: string | number, b: string | number): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

function cmpIds(a: string[], b: string[]): number {
  for (let i = 0; i < Math.min(a.length, b.length); i++) {
    const c = cmp(a[i]!, b[i]!);
    if (c !== 0) return c;
  }
  return a.length - b.length;
}

function findingOrder(a: Finding, b: Finding): number {
  return (
    BUCKET_ORDER.indexOf(a.bucket as (typeof BUCKET_ORDER)[number]) -
      BUCKET_ORDER.indexOf(b.bucket as (typeof BUCKET_ORDER)[number]) ||
    b.amount - a.amount ||
    cmp(a.type, b.type) ||
    cmpIds(a.statement_line_ids, b.statement_line_ids) ||
    cmpIds(a.ledger_line_ids, b.ledger_line_ids)
  );
}

function without<T extends { id: string }>(lines: T[], ids: Set<string>): T[] {
  return lines.filter((x) => !ids.has(x.id));
}

function byDateId(
  a: { doc_date: string; id: string },
  b: { doc_date: string; id: string },
): number {
  return cmp(a.doc_date, b.doc_date) || cmp(a.id, b.id);
}

export function reconcile(
  statementIn: StatementLine[],
  ledger: LedgerLine[],
  supplier: string,
  asAt: string,
  config: Config = DEFAULT_CONFIG,
  ledgerAsAt: string | null = null,
): ReconcileResult {
  const warnings: string[] = [];
  if (ledgerAsAt !== null && ledgerAsAt !== asAt) {
    warnings.push(
      `ledger as-at ${ledgerAsAt} does not equal statement as-at ${asAt}: ` +
        "the two balances describe different dates and differences may be " +
        "timing artefacts",
    );
  }

  const supplierLedger = ledger
    .filter((l) => l.supplier === supplier)
    .sort(byDateId);
  const statement = [...statementIn].sort(byDateId);
  const sById = new Map(statement.map((s) => [s.id, s]));
  const lById = new Map(supplierLedger.map((l) => [l.id, l]));

  let ledgerOpenTotal = 0;
  for (const l of supplierLedger) ledgerOpenTotal += l.open_amount;
  let statementTotal = 0;
  for (const s of statement) statementTotal += s.amount;

  // Pass 0 — duplicate scan, before any matching.
  const [dupFindings, excluded] = pass0Duplicates(
    supplierLedger, config.duplicateWindowDays, config.duplicateTightDays,
  );
  let remL = supplierLedger.filter((l) => !excluded.has(l.id));
  let remS = [...statement];
  const matches: Match[] = [];

  // Pass 1 — exact reference.
  const [m1, ms1, ml1] = pass1ExactRef(remS, remL);
  matches.push(...m1);
  remS = without(remS, ms1);
  remL = without(remL, ml1);

  // Pass 2 — normalised reference.
  const [m2, proposedRules, ms2, ml2] = pass2NormalisedRef(remS, remL, supplier);
  matches.push(...m2);
  remS = without(remS, ms2);
  remL = without(remL, ml2);

  // Pass 3 — ref matches, amounts differ: linked, not matched.
  const [links, cs, cl] = pass3RefAmountDiffer(remS, remL);
  remS = without(remS, cs);
  remL = without(remL, cl);

  // Pass 4 — amount + date window.
  const [m4, ms4, ml4] = pass4AmountDate(remS, remL, config.amountDateWindowDays);
  matches.push(...m4);
  remS = without(remS, ms4);
  remL = without(remL, ml4);

  // Pass 5 — bounded subset sums, both directions.
  const [m5, ms5, ml5] = pass5SubsetSums(remS, remL, config.subsetMaxSize);
  matches.push(...m5);
  remS = without(remS, ms5);
  remL = without(remL, ml5);

  // Pass 6 — classify residuals.
  let findings = [
    ...dupFindings,
    ...pass6Classify(remS, remL, links, matches, sById, lById, asAt, config.timingDays),
  ].sort(findingOrder);

  const bridge = buildBridge(findings, ledgerOpenTotal, statementTotal, sById, lById);

  let diagnostic: string | null = null;
  if (!bridge.ties_out) {
    let adjustmentTotal = 0;
    for (const a of bridge.adjustments) adjustmentTotal += a.amount;
    const gap = statementTotal - ledgerOpenTotal - adjustmentTotal;
    diagnostic =
      "bridge does not tie out: ledger open total plus adjustments misses " +
      `the statement total by ${gap} minor units; findings suppressed`;
    findings = [];
  }

  return {
    supplier,
    as_at: asAt,
    matches,
    findings,
    proposed_rules: proposedRules,
    bridge,
    warnings,
    diagnostic,
  };
}

File: src/tieout/engine.py
"""Reconciliation cascade orchestrator.

Order is a correctness requirement:
  Pass 0 duplicates -> 1 exact ref -> 2 normalised ref -> 3 ref-match/amount-differ
  -> 4 amount+date -> 5 subset sums -> 6 classify residuals -> bridge.

Pass 0 must run before any matching: if matching ran first, one of a
duplicate pair would be consumed by a match and the duplicate would become
undetectable.
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import date

from .bridge import build_bridge
from .models import BUCKET_ORDER, Finding, LedgerLine, ReconcileResult, StatementLine
from .passes import (
    pass0_duplicates, pass1_exact_ref, pass2_normalised_ref,
    pass3_ref_amount_differ, pass4_amount_date, pass5_subset_sums, pass6_classify,
)


@dataclass(frozen=True)
class Config:
    duplicate_window_days: int = 30
    duplicate_tight_days: int = 7
    amount_date_window_days: int = 5
    timing_days: int = 3
    subset_max_size: int = 6


def _without(lines, ids):
    return [x for x in lines if x.id not in ids]


def _finding_sort_key(f: Finding):
    return (BUCKET_ORDER.index(f.bucket), -f.amount, f.type,
            f.statement_line_ids, f.ledger_line_ids)


def reconcile(
    statement: list[StatementLine],
    ledger: list[LedgerLine],
    supplier: str,
    as_at: date,
    config: Config = Config(),
    ledger_as_at: date | None = None,
) -> ReconcileResult:
    warnings: list[str] = []
    if ledger_as_at is not None and ledger_as_at != as_at:
        warnings.append(
            f"ledger as-at {ledger_as_at.isoformat()} does not equal statement "
            f"as-at {as_at.isoformat()}: the two balances describe different "
            f"dates and differences may be timing artefacts"
        )

    supplier_ledger = sorted(
        (l for l in ledger if l.supplier == supplier), key=lambda l: (l.doc_date, l.id)
    )
    statement = sorted(statement, key=lambda s: (s.doc_date, s.id))
    s_by_id = {s.id: s for s in statement}
    l_by_id = {l.id: l for l in supplier_ledger}

    ledger_open_total = sum(l.open_amount for l in supplier_ledger)
    statement_total = sum(s.amount for s in statement)

    # Pass 0 — duplicate scan, before any matching.
    dup_findings, excluded = pass0_duplicates(
        supplier_ledger, config.duplicate_window_days, config.duplicate_tight_days
    )
    rem_l = [l for l in supplier_ledger if l.id not in excluded]
    rem_s = list(statement)
    matches = []

    # Pass 1 — exact reference.
    m1, ms, ml = pass1_exact_ref(rem_s, rem_l)
    matches += m1
    rem_s, rem_l = _without(rem_s, ms), _without(rem_l, ml)

    # Pass 2 — normalised reference.
    m2, proposed_rules, ms, ml = pass2_normalised_ref(rem_s, rem_l, supplier)
    matches += m2
    rem_s, rem_l = _without(rem_s, ms), _without(rem_l, ml)

    # Pass 3 — ref matches, amounts differ: linked, not matched.
    links, cs, cl = pass3_ref_amount_differ(rem_s, rem_l)
    rem_s, rem_l = _without(rem_s, cs), _without(rem_l, cl)

    # Pass 4 — amount + date window.
    m4, ms, ml = pass4_amount_date(rem_s, rem_l, config.amount_date_window_days)
    matches += m4
    rem_s, rem_l = _without(rem_s, ms), _without(rem_l, ml)

    # Pass 5 — bounded subset sums, both directions.
    m5, ms, ml = pass5_subset_sums(rem_s, rem_l, config.subset_max_size)
    matches += m5
    rem_s, rem_l = _without(rem_s, ms), _without(rem_l, ml)

    # Pass 6 — classify residuals.
    findings = dup_findings + pass6_classify(
        rem_s, rem_l, links, matches, s_by_id, l_by_id, as_at, config.timing_days
    )
    findings = sorted(findings, key=_finding_sort_key)

    bridge = build_bridge(findings, ledger_open_total, statement_total, s_by_id, l_by_id)

    diagnostic = None
    if not bridge.ties_out:
        gap = statement_total - ledger_open_total - sum(a.amount for a in bridge.adjustments)
        diagnostic = (
            "bridge does not tie out: ledger open total plus adjustments misses "
            f"the statement total by {gap} minor units; findings suppressed"
        )
        findings = []

    return ReconcileResult(
        supplier=supplier, as_at=as_at, matches=tuple(matches),
        findings=tuple(findings), proposed_rules=tuple(proposed_rules),
        bridge=bridge, warnings=tuple(warnings), diagnostic=diagnostic,
    )
